src/tools_for_rules.jl

```diff
@@ -0,0 +1,477 @@
+```jldoctest overlay-doctest
+julia> scale(x) = 2x
+scale (generic function with 1 method)
+
+julia> rule = Mooncake.build_rrule(Tuple{typeof(scale), Float64});
+
+julia> Mooncake.value_and_gradient!!(rule, scale, 5.0)
+(10.0, (NoTangent(), 2.0))
+```
+
+We can use `@mooncake_overlay` to change the definition which Mooncake.jl sees:
+```jldoctest overlay-doctest
+julia> Mooncake.@mooncake_overlay scale(x) = 3x
+
+julia> rule = Mooncake.build_rrule(Tuple{typeof(scale), Float64});
+
+julia> Mooncake.value_and_gradient!!(rule, scale, 5.0)
+(15.0, (NoTangent(), 3.0))
+```
+As can be seen from the output, the result of differentiating using Mooncake.jl has changed
+to reflect the overlay-ed definition of the method.
+
+Additionally, it is possible to use the usual multi-line syntax to declare an overlay:
+```jldoctest overlay-doctest
+julia> Mooncake.@mooncake_overlay function scale(x)
+           return 4x
+       end
+
+julia> rule = Mooncake.build_rrule(Tuple{typeof(scale), Float64});
+
+julia> Mooncake.value_and_gradient!!(rule, scale, 5.0)
+(20.0, (NoTangent(), 4.0))
+```
+"""
+macro mooncake_overlay(method_expr)
+    def = splitdef(method_expr)
+    def[:name] = Expr(:overlay, :(Mooncake.mooncake_method_table), def[:name])
+    return esc(combinedef(def))
+end
+
+#
+# Functionality supporting @zero_adjoint
+#
+
+"""
+    zero_adjoint(f::CoDual, x::Vararg{CoDual, N}) where {N}
+
+Utility functionality for constructing `rrule!!`s for functions which produce adjoints which
+always return zero.
+
+NOTE: you should only make use of this function if you cannot make use of the
+[`@zero_adjoint`](@ref) macro.
+
+You make use of this functionality by writing a method of `Mooncake.rrule!!`, and
+passing all of its arguments (including the function itself) to this function. For example:
+```jldoctest
+julia> import Mooncake: zero_adjoint, DefaultCtx, zero_fcodual, rrule!!, is_primitive, CoDual
+
+julia> foo(x::Vararg{Int}) = 5
+foo (generic function with 1 method)
+
+julia> is_primitive(::Type{DefaultCtx}, ::Type{<:Tuple{typeof(foo), Vararg{Int}}}) = true;
+
+julia> rrule!!(f::CoDual{typeof(foo)}, x::Vararg{CoDual{Int}}) = zero_adjoint(f, x...);
+
+julia> rrule!!(zero_fcodual(foo), zero_fcodual(3), zero_fcodual(2))[2](NoRData())
+(NoRData(), NoRData(), NoRData())
+```
+
+WARNING: this is only correct if the output of `primal(f)(map(primal, x)...)` does not alias
+anything in `f` or `x`. This is always the case if the result is a bits type, but more care
+may be required if it is not.
+```
+"""
+@inline function zero_adjoint(f::CoDual, x::Vararg{CoDual, N}) where {N}
+    return zero_fcodual(primal(f)(map(primal, x)...)), NoPullback(f, x...)
+end
+
+"""
+    @zero_adjoint ctx sig
+
+Defines `is_primitive(context_type, sig) = true`, and defines a method of
+`Mooncake.rrule!!` which returns zero for all inputs.
+Users of ChainRules.jl should be familiar with this functionality -- it is morally the same
+as `ChainRulesCore.@non_differentiable`.
```

Comment:
The name `@non_differentiable` implies that the function cannot be differentiated / does not have a derivative (or, therefore, an adjoint). I actually think `@non_differentiable` is quite an unfortunate name in hindsight, because it suggests that a given function _can't_ be differentiated, which isn't at all what the purpose of the macro is. The macro actually _means_ "the derivative / adjoint is zero for all inputs".

Conversely, the name `zero_adjoint` aligns perfectly with what we mean.